lib/tasks/multitenant/tenants.rake

```diff
@@ -77,6 +77,40 @@ namespace :multitenant do
       Rails.logger.error "Inconsistent tenant_ids for:\n#{inconsistent.map {_1.join(" ")}.join("\n")}"
     end
 
+    # NOTE: this task may cycle forever in case some unforeseen bug causes more tenants than
+    #       the number of desired concurrency to fail being deleted
+    desc 'Schedule stale tenants background deletion.'
+    task :stale_throttled_delete, %i[concurrency days_since_disabled iteration_wait] => :environment do |_task, args|
+      require "progress_counter"
+
+      # Not using Account::States::PERIOD_BEFORE_DELETION because customers using this task
+      # will probably have FindAndDeleteScheduledAccountsWorker disabled. To err on the safe side
+      # we only delete ancient stuff by default.
+      args.with_defaults(:concurrency => 3, :days_since_disabled => 30*6, :iteration_wait => 60)
+      target_concurrency = Integer(args.concurrency)
+      since = Integer(args.days_since_disabled).days.ago
+      iteration_wait = Integer(args.iteration_wait)
+
+      deletion_scope = ->{ Account.tenants.deleted_since(since) }
+
+      progress = ProgressCounter.new(deletion_scope.call.count)
+      loop do
+        deletions = scheduled_or_running_background_deletions
+        to_schedule = target_concurrency - deletions.count
+        unless to_schedule.zero?
```

Comment:
What if the task is run when there are already some deletions in the queue? `to_schedule` maybe become a negative number, which doesn't make sense in `limit`.
Maybe this condition should be `if to_schedule > 0` ?

Reply:
I have envisioned this scenario, yet this check was not updated. Good catch.